### ISSUE TITLE
[Discussion] Add a C Library to GluonNLP

### ISSUE THREAD

**Issue by sxjscience** [dmlc/gluon-nlp#147]
As reported in https://github.com/dmlc/gluon-nlp/pull/136, some functions can be greatly accelerated using C implementations. Would it be worthwhile to add a C library to hold these functions? In fact, FastText are generally implemented in C++  (https://github.com/facebookresearch/fastText/tree/master/src) and the bleu score in fairseq is also implemented in C (https://github.com/pytorch/fairseq/tree/master/fairseq/clib/libbleu).

**Comment by szha** [comment]
Indeed. On the other hand, the concern around having C library is it complicates the testing setup here. We might not have the luxury of the testing suites of many different environments in mxnet.

**Comment by leezu** [comment]
For most processing steps numba is sufficient to speed up. Unfortunately numba does not (yet) support string handling, which in a NLP library is quite central.. So it can only be used after a dataset has been numericalized / coded. Consequently I think it's useful to add it, though I'm not sure how it affects testing (ie. requires testing on Windows?)

**Comment by szha** [comment]
One approach that @eric-haibin-lin took was to include a make file and the c code in the example, and ask users to compile it when using.

**Comment by sxjscience** [comment]
Yes. We can use CMake to also benefit the windows users.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by leezu** [comment]
While numba has been quite helpful so far, I did run into limitations as the numba jitclass support is not yet sufficiently optimized and would becomes the bottleneck. I guess this will be addressed by numba at some point, but for now that does not help us.
For easily supporting Cython, we can choose the Pure Python mode, allowing any Cython code we write to be interpreted if Cython is not installed given that we include a copy of Cython Shadow.py in GluonNLP.

Also see http://docs.cython.org/en/latest/src/tutorial/pure.html and https://github.com/cython/cython/blob/0.29.x/Cython/Shadow.py

**Comment by sxjscience** [comment]
Cython + 1, We should not only rely on numba. Cython allows more customization. Another solution is use CPython directly. However it will require compilation. (Cython requires compilation too)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by leezu** [comment]
I would argue we can simply add a *requirement* on Cython in the following sense:

- python distutils (`setup.py`) transparently falls back to transpiled C code if the user does not have cython. We therefore only require the presence of a C compiler as long as we distribute the platform independent transpiled C code. The C code is obtained by calling `cythonize` on the `.pyx` files. This can be automated easily in `setup.py`
- We can offer precompiled version on pip for same platforms as mxnet. If the user does not have a C compiler, they may use the version from pip. If it is not available for their platform, then they cannot use our package anyways due to missing mxnet. Or they compile mxnet themselves, in which case they should have a C compiler..

**Comment by sxjscience** [comment]
This solution LGTM

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by sxjscience** [comment]
@leezu I think we can revisit about whether it's worthwhile to add a C Library to the package.
